Mathlib/Combinatorics/SetFamily/KruskalKatona.lean

```diff
@@ -0,0 +1,344 @@
+    · rw [← compress_sdiff_sdiff A B] at hAB hBA
+      cases hBA.not_lt $ toColex_compress_lt_toColex hlt hAB
+    · exact (disjoint_right.1 disj (max'_mem _ hU) $ heq.symm ▸ max'_mem _ _).elim
+    · assumption
+  refine hB ?_
+  rw [← (h₂ _ _ ⟨disj, card_sdiff_comm hAB'.symm, hV, hU, smaller⟩).eq]
+  exact mem_compression.2 (Or.inr ⟨hB, A, hA, compress_sdiff_sdiff _ _⟩)
+
+attribute [-instance] Fintype.decidableForallFintype
+
+-- TODO: There's currently a diamond
+-- import Mathlib.Order.Fin.Basic
+-- example (n : ℕ) : instDecidableEqFin n = instDecidableEq_mathlib := rfl
+attribute [-instance] instDecidableEqFin
+
+/-- This measures roughly how compressed the family is.
+
+Note that this does depend on the order of the ground set, unlike the Kruskal-Katona theorem itself
+(although `kruskal_katona` currently is stated in an order-dependent manner). -/
+private def familyMeasure (𝒜 : Finset (Finset (Fin n))) : ℕ := ∑ A in 𝒜, ∑ a in A, 2 ^ (a : ℕ)
+
+/-- Applying a compression strictly decreases the measure. This helps show that "compress until we
+can't any more" is a terminating process. -/
+lemma familyMeasure_compression_lt_familyMeasure {U V : Finset (Fin n)} {hU : U.Nonempty}
+    {hV : V.Nonempty} (h : max' U hU < max' V hV) {𝒜 : Finset (Finset (Fin n))} (a : 𝓒 U V 𝒜 ≠ 𝒜) :
+    familyMeasure (𝓒 U V 𝒜) < familyMeasure 𝒜 := by
+  rw [compression] at a ⊢
+  have q : ∀ Q ∈ 𝒜.filter fun A ↦ compress U V A ∉ 𝒜, compress U V Q ≠ Q := by
+    simp_rw [mem_filter]
+    intro Q hQ h
+    rw [h] at hQ
+    exact hQ.2 hQ.1
+  have uA : (𝒜.filter fun A => compress U V A ∈ 𝒜) ∪ 𝒜.filter (fun A ↦ compress U V A ∉ 𝒜) = 𝒜 :=
+    filter_union_filter_neg_eq _ _
+  have ne₂ : (𝒜.filter fun A ↦ compress U V A ∉ 𝒜).Nonempty := by
+    refine nonempty_iff_ne_empty.2 fun z ↦ a ?_
+    rw [filter_image, z, image_empty, union_empty]
+    rwa [z, union_empty] at uA
+  rw [familyMeasure, familyMeasure, sum_union compress_disjoint]
+  conv_rhs => rw [← uA]
+  rw [sum_union (disjoint_filter_filter_neg _ _ _), add_lt_add_iff_left, filter_image,
+    sum_image compress_injOn]
+  refine sum_lt_sum_of_nonempty ne₂ fun A hA ↦ ?_
+  simp_rw [← sum_image Fin.val_injective.injOn]
+  rw [geomSum_lt_geomSum_iff_toColex_lt_toColex le_rfl,
+    toColex_image_lt_toColex_image Fin.val_strictMono]
+  exact toColex_compress_lt_toColex h $ q _ hA
+
+/-- The main Kruskal-Katona helper: use induction with our measure to keep compressing until
+we can't any more, which gives a set family which is fully compressed and has the nice properties we
+want. -/
+private lemma kruskal_katona_helper {r : ℕ} (𝒜 : Finset (Finset (Fin n)))
+    (h : (𝒜 : Set (Finset (Fin n))).Sized r) :
+    ∃ ℬ : Finset (Finset (Fin n)), (∂ ℬ).card ≤ (∂ 𝒜).card ∧ 𝒜.card = ℬ.card ∧
+      (ℬ : Set (Finset (Fin n))).Sized r ∧ ∀ U V, UsefulCompression U V → IsCompressed U V ℬ := by
+  classical
+  revert h
+  apply WellFounded.recursion (InvImage.wf familyMeasure wellFounded_lt) 𝒜
```

Comment:
tempting to try strong induction on family measure here... but can leave it for now if you want

Reply:
That was easy :)